Adding generation of hunter tents

Cheaterpaul opened the pull request:
* generate 2 types of hunter tents
* advanced hunter tents
* change registryname of BasicHunterEntity from vampire_hunter to hunter
* introduce faction biomes
* package private packaging

Cheaterpaul committed:
removed unnecessary code

Cheaterpaul committed:
first hunter camp functionality

still won't work properly

Cheaterpaul committed:
introduce faction biomes

Cheaterpaul committed:
modifies checkForVampireBiome command

Cheaterpaul committed:
properly generated camps

Cheaterpaul commented:
@maxanier please test the formation or etc. about the camp and inform me about everything you dont like

Cheaterpaul committed:
package private packaging

maxanier commented on src/main/java/de/teamlapen/vampirism/VampirismMod.java:
```diff
@@ -243,6 +243,7 @@ private void loadComplete(final FMLLoadCompleteEvent event) {
     }
 
     private void processIMC(final InterModProcessEvent event) {
+        registryManager.onInitStep(IInitListener.Step.PROCESS_IMC, event);
```
Pls swap with setupAPI2, unless there it is required beforehand

maxanier commented on src/main/java/de/teamlapen/vampirism/config/VampirismConfig.java:
```diff
@@ -272,6 +273,7 @@ public static void onReload(final ModConfig.ConfigReloading configEvent) {
             versionCheck = builder.comment("Check for new versions of Vampirism on startup").define("versionCheck", true);
             collectStats = builder.comment("Send mod version, MC version and mod count to mod author").define("collectStats", true);
             disableVampireForest = builder.comment("Disable vampire forest generation").define("disableVampireForest", false);
+            disableHunterTentGen = builder.comment("Disable hunter camp generation").define("disableHunterTentGen", false);
```
Not sure why I put disableVampireForest here. Both options should go into the server config

maxanier commented on src/main/java/de/teamlapen/vampirism/core/ModBiomes.java:
```diff
@@ -34,9 +35,11 @@ static void addBiome() {
         for (Biome biome : ForgeRegistries.BIOMES.getValues()) {
-            if (new ResourceLocation("the_end").equals(biome.getRegistryName()) || new ResourceLocation("nether").equals(biome.getRegistryName()))
+            if (nonHunterCampBiomeCategories.contains(biome.getCategory())) continue;
+            if (biome instanceof IFactionBiome && VReference.VAMPIRE_FACTION.equals(((IFactionBiome) biome).getFaction()))
```
I think it would be better to do !HUNTER.equals(biome.getFaction)

maxanier commented on src/main/java/de/teamlapen/vampirism/core/ModBiomes.java:
```diff
@@ -34,9 +35,11 @@ static void addBiome() {
 
-    static void registerFeatures() {
+    static void addFeatures() {
+        List nonHunterCampBiomeCategories = Lists.newArrayList(Biome.Category.OCEAN, Biome.Category.THEEND, Biome.Category.NETHER, Biome.Category.ICY, Biome.Category.BEACH, Biome.Category.RIVER);
```
We need some kind of API access for this, so e.g. VampirismIntegrations can exclude certain BiomeOPlenties biomes

maxanier commented on src/main/java/de/teamlapen/vampirism/world/gen/biome/VampireBiome.java:
```diff
@@ -0,0 +1,18 @@
+import de.teamlapen.vampirism.api.entity.factions.IFaction;
+import de.teamlapen.vampirism.api.world.IFactionBiome;
+
+public abstract class VampireBiome extends VampirismBiome implements IFactionBiome {
```
Not really necessary to add another class that has only one subclass (for the forseable future) and does very little, but ok

maxanier commented on src/main/java/de/teamlapen/vampirism/world/gen/biome/VampirismBiome.java:
```diff
@@ -0,0 +1,12 @@
+import de.teamlapen.vampirism.util.REFERENCE;
+import net.minecraft.world.biome.Biome;
+
+public abstract class VampirismBiome extends Biome {
```
Same.

maxanier commented on src/main/java/de/teamlapen/vampirism/world/loot/LootHandler.java:
```diff
@@ -20,13 +21,6 @@
 
     public static final ResourceLocation STRUCTURE_VAMPIRE_DUNGEON = register("vampire_dungeon");
-    public static final ResourceLocation STRUCTURE_VILLAGE_TRAINER = register("village_trainer");
-    public static final ResourceLocation BASIC_VAMPIRE = register("entities/basic_vampire");
```
I assume this is not necessary anymore?

maxanier replied:
Could we avoid duplicate (imob and non imob) loot tables if we kept this?

Cheaterpaul replied:
ok, reimplement it for imob and non imob entities

Cheaterpaul replied:
see ModLootTables

in the next time i will try to move the stuff from LootHandler to ModLootTables

maxanier commented on src/main/java/de/teamlapen/vampirism/world/gen/structures/village/VillagePieceTotem.java:
```diff
@@ -1,137 +0,0 @@
-//        this.setBlockState(worldIn, ModBlocks.totem_base.getDefaultState(), 1, 0, 1, structureBoundingBoxIn);
-//        this.setBlockState(worldIn, ModBlocks.totem_top.getDefaultState(), 1, 1, 1, structureBoundingBoxIn);
-//
-//        BlockPos blockpos = new BlockPos(this.getXWithOffset(1, 1), this.getYWithOffset(1), this.getZWithOffset(1, 1));
```
We don't have a replacement yet, right?

Cheaterpaul replied:
no, waiting for forge

maxanier reviewed (commented):
Hey, happy to hear that it is working now. I haven't had the time to test it or have a closer look at the generation, just a few things I noticed when scanning this

maxanier commented:
Looks very good. Works very well even in the mountains. I have seen a few fire places or crafting tables without any tents nearby, but that is ok I guess.

I would remove swamp and jungle from the biome list and make sure the camps do not generate near or even in villages.

Last but not least we need two different types of camps. One smaller variant (1-2 tents) and one bigger (2-4, maybe even a weapon table) which spawns a single advanced hunter.

Cheaterpaul committed:
requested changes

- move method call
- move world configs
- remove duplicated loot table for imob & non imob entities
- remove unnecessary code
- add WorldGenManager fro VampirismAPI
- changed registryname of vampirism:vampire_hunter to vampirism:hunter for consistency

maxanier reviewed (commented):
Changing the id of hunters is rather severe change, that probably should have been done before releasing the first alpha as it causes missing id errors.
Pls have a look at the MissingMappings event to create a remapping from the old id to the new one

Cheaterpaul committed:
introduce 2 types of camps

- fix mapping of hunter entity
- spawning of advanced still not working

maxanier commented on src/main/java/de/teamlapen/vampirism/config/BalanceConfig.java:
```diff
@@ -105,6 +107,8 @@
         builder.push("general");
 
+        hunterTentDistance = builder.comment("dont set hunterTentDistance <= hunterTentSeperation").defineInRange("hunterTentDistance", 8, 2, Integer.MAX_VALUE);
+        hunterTentSeperation = builder.comment("dont set hunterTentDistance <= hunterTentSeperation").defineInRange("hunterTentSeperation", 3, 1, Integer.MAX_VALUE);
```
What happens if one does set the value lower, crash or something else?
If it is something fatal, we could throw an exception somewhere (e.g. `#getBiomeFeatureDistance`) with a meaningful error message

maxanier replied:
Alternatively, we could just fall back to a valid value, if the configuration is invalid

maxanier commented on src/main/java/de/teamlapen/vampirism/config/BalanceConfig.java:
```diff
@@ -7,6 +7,8 @@
 
     //GENERAL
+    public final ForgeConfigSpec.IntValue hunterTentDistance;
+    public final ForgeConfigSpec.IntValue hunterTentSeperation;
```
separation

Cheaterpaul commented on src/main/java/de/teamlapen/vampirism/world/gen/structures/huntercamp/HunterCampStructure.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    protected int getBiomeFeatureDistance(ChunkGenerator<?> chunkGeneratorIn) {
```
```suggestion
    @Override
    protected int getBiomeFeatureDistance(ChunkGenerator<?> chunkGeneratorIn) {
        if(VampirismConfig.BALANCE.hunterTentDistance.get() <= VampirismConfig.BALANCE.hunterTentSeparation.get()){
            LogManager.getLogger(BalanceConfig.class).warn("config value 'hunterTentDistance' is not set greater than 'hunterTentSeparation'. 'hunterTentDistance' increased");
            VampirismConfig.BALANCE.hunterTentDistance.set(VampirismConfig.BALANCE.hunterTentSeparation.get()+1);
        }
        return VampirismConfig.BALANCE.hunterTentDistance.get();
    }
```

maxanier replied:
:ok_hand:

Cheaterpaul committed:
hunter camp config failsafe

Cheaterpaul committed:
Advanced hunter spawner in tent tileentity

Cheaterpaul committed:
Changed Biome.Category to BiomeDictionary.Type for the ignored biomes of structures

maxanier commented:
Travis build fails, BiomeDictonary cannot be found

maxanier commented:
Excellent :ok_hand:

maxanier commented:
Excellent

Cheaterpaul committed:
Adding missing import
